proposals/ytt/005-semver/README.md

```diff
@@ -0,0 +1,217 @@
+---
+title: "Semver support in the standard library"
+authors: [ "Max Brauer <mbrauer@vmware.com>" ]
+status: "draft"
+approvers: [ "Dmitriy Kalinin <dkalinin@vmware.com>", "Join Ryan <ryanjo@vmware.com>" ]
+---
+
+# Semver
+
+## Problem Statement
+
+Configuration authors frequently encounter strings which denote the version of a
+software. [Semantic versioning](https://semver.org/) is a widely adopted scheme for versioning
+software. In this proposal the term "semver" is used both for _the_ scheme itself as well _an_ instance of a version.
+
+A semantic version is a string, e.g. `3.2.12`, `0.23.0-beta`, `1.99.2-next+build.5`, which contains structured
+information about the version. Configuration authors may want
+to read a version's constituents, mutate them and compare different versions. For example, when authoring `ytt`
+templates:
+
+* I want to assert that a given string is a well-formed semver.
+* I want to change a semver's representation from string another to data structure, e.g. object or array.
+* I want to compare semvers, e.g. expect a given version to be larger than a known version.
+* I want to (de,in)crement the major, minor and/or patch version of a semver.
+* I want to add, change or remove a semver's prerelease identifier or metadata.
+
+Starlark's standard library provides the means to achieve this. However, it sets a high bar and burdens configuration
+authors to write boilerplate code for common, industry standard operations.
+
+## Terminology / Concepts
+
+A semantic version encodes different properties about the version in a string of the
+form `<major>.<minor>.<patch>(-<pre-release>)(+<metadata>)`.
+
+Semantic versions are strictly sortable.
```

Comment:
Is there exactly one way to sort semantic versions (specially when including pre-release and/or metadata)?

Are there cases where users might want to modify that sort order?

Reply:
The semver spec defines [precedence](https://semver.org/#spec-item-11). In that sense, yes there's exactly one way to sort them. `metadata` is excluded from precedence as per spec.

I think this is better:
```suggestion
Semantic versions can be sorted based on [precedence](https://semver.org/#spec-item-11).
```

> Are there cases where users might want to modify that sort order?
I would argue that if you need to change the notion of precedence, then it's not semver any more. However, if one needs to change the precedence then should be able to creating their own `#@ def sort_semver_but_different(...`. However, I find it very unlikely to be a realistic requirement. I may be wrong.